File: core/src/main/java/pcb/uwu/amount/quantity/Length.java
package pcb.uwu.amount.quantity;

import pcb.uwu.amount.composite.fundamental.Area;
import pcb.uwu.amount.composite.mechanics.Speed;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.composite.fundamental.AreaUnit;
import pcb.uwu.units.composite.mechanics.SpeedUnit;
import pcb.uwu.units.quantity.LengthUnit;
import pcb.uwu.units.quantity.TimeUnit;

import java.math.BigDecimal;
import java.math.MathContext;

public class Length extends CompositeUnitAmount<LengthUnit> {

	// region constructors

	public Length(Number value, LengthUnit unit) {
		super(value, unit);
	}

	public Length(Number value, Magnitude magnitude, LengthUnit unit) {
		super(value, magnitude, unit);
	}

	public Length(String value, LengthUnit unit) {
		super(value, unit);
	}

	public Length(String value, Magnitude magnitude, LengthUnit unit) {
		super(value, magnitude, unit);
	}

	public Length(BigDecimal value, LengthUnit unit) {
		super(value, unit);
	}

	public Length(BigDecimal value, Magnitude magnitude, LengthUnit unit) {
		super(value, magnitude, unit);
	}

	public Length(BigDecimalAmount amount, LengthUnit unit) {
		super(amount, unit);
	}

	public Length(BigDecimalAmount amount, Magnitude magnitude, LengthUnit unit) {
		super(amount, magnitude, unit);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Length plus(UnitAmount<LengthUnit> other, MathContext mathContext) {
		return new Length(plusAmount(other, mathContext), getUnit());
	}

	@Override
	public Length minus(UnitAmount<LengthUnit> other, MathContext mathContext) {
		return new Length(minusAmount(other, mathContext), getUnit());
	}

	@Override
	public Length multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Length(multipliedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Length dividedBy(BigDecimal other, MathContext mathContext) {
		return new Length(dividedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Length convertTo(LengthUnit unit) {
		return new Length(getAmountIn(unit), unit);
	}

	// endregion

	// region composition

	public Speed dividedBy(Time time, MathContext mathContext) {
		BigDecimalAmount amount = getAmount()
				.dividedBy(time.getAmount().getValue(), mathContext);
		SpeedUnit unit = new SpeedUnit(getUnit(), time.getUnit());

		return new Speed(amount, unit);
	}

	public Time dividedBy(Speed speed, MathContext mathContext) {
		TimeUnit speedTimeUnit = speed.getUnit().getUnitCounter().findUnit(TimeUnit.class);
		LengthUnit speedLengthUnit = speed.getUnit().getUnitCounter().findUnit(LengthUnit.class);

		BigDecimalAmount amount = getAmount()
				.dividedBy(speed.getAmountIn(new SpeedUnit(speedLengthUnit, speedTimeUnit)).getValue(), mathContext);

		return new Time(amount, speedTimeUnit);
	}

	public Area multipliedBy(Length length, MathContext mathContext) {
		return new Area(
				getAmount().multipliedBy(length.getAmount(), mathContext),
				new AreaUnit(getUnit(), length.getUnit()));
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/amount/base/Length.kt
package pcb.uwu.amount.base

import pcb.uwu.amount.derived.geometry.Area
import pcb.uwu.amount.derived.mechanics.Force
import pcb.uwu.amount.derived.mechanics.Speed
import pcb.uwu.amount.derived.thermodynamics.Energy
import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.LengthUnit
import pcb.uwu.unit.base.TimeUnit
import pcb.uwu.unit.derived.geometry.AreaUnit
import pcb.uwu.unit.derived.mechanics.SpeedUnit
import pcb.uwu.unit.derived.termodynamics.EnergyUnit
import pcb.uwu.util.UnitAmountUtils

open class Length : CompositeUnitAmount<LengthUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: LengthUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: LengthUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(length: UnitAmount<LengthUnit>) =
        Length(amount = this.amount + (length to this.unit).amount,
               unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(length: UnitAmount<LengthUnit>) =
        Length(amount = this.amount - (length to this.unit).amount,
               unit = this.unit)

    override fun times(scalar: Number) =
        Length(amount = this.amount * scalar,
               unit = this.unit)

    override fun div(scalar: Number) =
        Length(amount = this.amount / scalar,
               unit = this.unit)

    override fun to(unit: LengthUnit) =
        Length(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
               unit = unit)

    // endregion

    // region composition

    open operator fun div(time: Time): Speed =
        Speed(amount = this.amount / time.amount,
              unit = SpeedUnit(this.unit, time.unit))

    open operator fun div(speed: Speed) =
        Time(amount = super.div(speed).amount,
             unit = speed.unit.unitCounter.findUnit(TimeUnit::class)!!)

    open operator fun times(length: Length) =
        Area(amount = amount * length.amount,
             unit = AreaUnit(this.unit, length.unit))

    open operator fun times(force: Force) =
        Energy(amount = amount * force.amount,
               unit = EnergyUnit(lengthUnit = this.unit, forceUnit = force.unit))

    // endregion
}
